<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Empirical Bayes non-negative matrix factorization for single-cell RNA-seq data • flashier</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Empirical Bayes non-negative matrix factorization for single-cell RNA-seq data">
<meta property="og:description" content="flashier">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">flashier</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.0.53</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">Home</a>
</li>
<li>
  <a href="../articles/index.html">Vignettes</a>
</li>
<li>
  <a href="../reference/index.html">Functions</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/willwerscheid/flashier" class="external-link">Source</a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Empirical Bayes non-negative matrix
factorization for single-cell RNA-seq data</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/willwerscheid/flashier/blob/HEAD/vignettes/flashier_single_cell.Rmd" class="external-link"><code>vignettes/flashier_single_cell.Rmd</code></a></small>
      <div class="hidden name"><code>flashier_single_cell.Rmd</code></div>

    </div>

    
    
<p>The aim of this vignette is show how <code>flashier</code> can be
used to perform a non-negative matrix factorization (NMF) analysis of
single-cell RNA-seq data. This vignette is modeled after the <a href="https://stephenslab.github.io/fastTopics/articles/single_cell_rnaseq_basic.html" class="external-link">fastTopics
vignette on analyzing single-cell data</a>.</p>
<p>We begin by loading the required packages. We also set the seed so
that results can be fully reproduced.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/willwerscheid/flashier" class="external-link">flashier</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://Matrix.R-forge.R-project.org" class="external-link">Matrix</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://stephenslab.github.io/fastTopics/" class="external-link">fastTopics</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="preparing-the-single-cell-data-for-flashier">Preparing the single-cell data for flashier<a class="anchor" aria-label="anchor" href="#preparing-the-single-cell-data-for-flashier"></a>
</h2>
<p>The single-cell RNA-seq data (we use the <code>pbmc_facs</code> data
from the <code>fastTopics</code> package) are unique molecular
identifier (UMI) counts stored as an <span class="math inline">\(n
\times p\)</span> sparse matrix, where <span class="math inline">\(n\)</span> is the number of cells and <span class="math inline">\(p\)</span> is the number of genes:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"pbmc_facs"</span><span class="op">)</span></span>
<span><span class="va">counts</span> <span class="op">&lt;-</span> <span class="va">pbmc_facs</span><span class="op">$</span><span class="va">counts</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/make.names.html" class="external-link">make.names</a></span><span class="op">(</span><span class="va">pbmc_facs</span><span class="op">$</span><span class="va">genes</span><span class="op">$</span><span class="va">symbol</span>, unique <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1]  3774 16791</span></span></code></pre></div>
<p>(Note that other R packages, for example <a href="https://github.com/satijalab/seurat" class="external-link">Seurat</a>, use the
convention that rows are genes and columns are cells.)</p>
<p>Since the <code>flashier</code> model, like other linear-model-based
methods (e.g., principal components analysis), was not designed for
count data, it is recommended to first transform the data in a way that
makes them more suitable. A widely-used approach is to divide the counts
by a “size factor,” add a “pseudocount,” then take the log. We call this
transformation the “shifted logarithm,” following <a href="https://doi.org/10.1038/s41592-023-01814-1" class="external-link">this paper</a>, and we
refer to the transformed counts as “shifted log counts”.</p>
<p>In practice, the shifted log counts are computed as follows: we first
divide the counts by <span class="math inline">\(\alpha s_i\)</span>,
where <span class="math inline">\(s_i\)</span> is the size factor for
cell <span class="math inline">\(i\)</span>, and <span class="math inline">\(\alpha\)</span> is the pseudocount. Done this way,
the shifted log counts maintain sparsity in the data; that is, if the
original count is zero, then the shifted log count is also zero.</p>
<p>More formally, the transformed counts <span class="math inline">\(y_{ij}\)</span> obtained from the original counts
<span class="math inline">\(x_{ij}\)</span> are <span class="math display">\[
y_{ij} = \log\bigg(1 + \frac{x_{ij}}{\alpha s_i}\bigg).
\]</span> In our analysis below, we use library-size normalization; that
is, we set <span class="math display">\[
s_i =
\frac{\text{total count for cell} \; i}
     {\text{average across all cells}} =
\frac{\sum_{j=1}^p x_{ij}}
     {\frac{1}{n}\sum_{i=1}^n \sum_{j=1}^p x_{ij}},
\]</span> and we set the pseudocount <span class="math inline">\(\alpha\)</span> to 1.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">a</span> <span class="op">&lt;-</span> <span class="fl">1</span></span>
<span><span class="va">size_factors</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span></span>
<span><span class="va">size_factors</span> <span class="op">&lt;-</span> <span class="va">size_factors</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">size_factors</span><span class="op">)</span></span>
<span><span class="va">shifted_log_counts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log1p</a></span><span class="op">(</span><span class="va">counts</span> <span class="op">/</span> <span class="op">(</span><span class="va">a</span> <span class="op">*</span> <span class="va">size_factors</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Note that, since the counts are sparse, we could have performed this
transformation more efficiently using, for example, the
<code>mapSparse</code> function from the <a href="https://github.com/david-cortes/MatrixExtra" class="external-link">MatrixExtra</a>
package.</p>
</div>
<div class="section level2">
<h2 id="variance-regularization">Variance regularization<a class="anchor" aria-label="anchor" href="#variance-regularization"></a>
</h2>
<p>Before we fit the model, one issue we need to confront is that
<code>flashier</code> may automatically estimate the variances to be too
small, which can especially be an issue for genes with low expression.
We can avoid this issue by setting a sensible lower bound on the
variance estimates.</p>
<!-- When running flashier, we need to place some regularization on the -->
<!-- variance estimates. Without regularization, it is possible for some -->
<!-- variance estimates to be very small, which can lead to overfitting. -->
<p>We use the following rule of thumb: estimate the standard deviation
of the transformed data for a Poisson random variable with rate <span class="math inline">\(\mu = 1 / n\)</span>, where <span class="math inline">\(n\)</span> is the number of samples. This standard
deviation corresponds to a gene for which we would expect to observe a
single count across all <span class="math inline">\(n\)</span> cells, so
it can serve as a reasonable lower bound to prevent variance estimates
from getting too small.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">n</span>  <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span></span>
<span><span class="va">x</span>  <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/Poisson.html" class="external-link">rpois</a></span><span class="op">(</span><span class="fl">1e7</span>, <span class="fl">1</span><span class="op">/</span><span class="va">n</span><span class="op">)</span></span>
<span><span class="va">s1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/sd.html" class="external-link">sd</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="va">x</span> <span class="op">+</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="fit-a-flashier-model">Fit a flashier model<a class="anchor" aria-label="anchor" href="#fit-a-flashier-model"></a>
</h2>
<p>Now we can call flash to fit an NMF model to the transformed
data:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/flash.html">flash</a></span><span class="op">(</span><span class="va">shifted_log_counts</span>, </span>
<span>             ebnm_fn <span class="op">=</span> <span class="va">ebnm_point_exponential</span>,</span>
<span>             var_type <span class="op">=</span> <span class="fl">2</span>, </span>
<span>             greedy_Kmax <span class="op">=</span> <span class="fl">8</span>, </span>
<span>             S <span class="op">=</span> <span class="va">s1</span>,</span>
<span>             backfit <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="co">#&gt; Adding factor 1 to flash object...</span></span>
<span><span class="co">#&gt; Adding factor 2 to flash object...</span></span>
<span><span class="co">#&gt; Adding factor 3 to flash object...</span></span>
<span><span class="co">#&gt; Adding factor 4 to flash object...</span></span>
<span><span class="co">#&gt; Adding factor 5 to flash object...</span></span>
<span><span class="co">#&gt; Adding factor 6 to flash object...</span></span>
<span><span class="co">#&gt; Adding factor 7 to flash object...</span></span>
<span><span class="co">#&gt; Adding factor 8 to flash object...</span></span>
<span><span class="co">#&gt; Wrapping up...</span></span>
<span><span class="co">#&gt; Done.</span></span>
<span><span class="co">#&gt; Nullchecking 8 factors...</span></span>
<span><span class="co">#&gt; Done.</span></span></code></pre></div>
<p>A few notes about this flash call:</p>
<ul>
<li><p><code>ebnm_fn = ebnm_point_exponential</code> forces both the
<span class="math inline">\({\bf L}\)</span> and <span class="math inline">\({\bf F}\)</span> matrices in flashier to be
non-negative, so this call will generate a non-negative matrix
factorization of the (transformed) counts matrix. Other types of matrix
factorizations can be produced with different choices of the
<code>ebnm_fn</code> argument; this is explained in detail in the <a href="https://willwerscheid.github.io/flashier/articles/flashier_intro.html" class="external-link">Introduction
to flashier vignette</a>. See also the notes at the bottom of this
vignette.</p></li>
<li><p><code>var_type = 2</code> means that we estimate column-wise
(here, gene-wise) variances; that is, we estimate a different variance
for each column (gene).</p></li>
<li><p><code>greedy_Kmax = 8</code> forces flashier to fit no more than
8 factors. <strong>In practice we recommend using a larger
value,</strong> but to keep the example simple and short we set it to 8
here.</p></li>
<li><p><code>backfit = FALSE</code> skips the backfitting step.
Backfitting can often greatly improve the fit, and is generally
recommended for better results. (And indeed, backfitting noticeably
improves the fit for these data.) On the other hand, backfitting can
sometimes take a long time to complete, so have omitted the backfitting
step here to reduce the computational effort involved.</p></li>
</ul>
<p>The <code>plot</code> method with default settings produces a “scree
plot” showing the proportion of variance explained by each of the 8
factors:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">fit</span><span class="op">)</span></span></code></pre></div>
<p><img src="flashier_single_cell_files/figure-html/scree-plot-1.png" width="300" style="display: block; margin: auto;"></p>
<p>Factors 7 and 8 explain a very small proportion of variation (and
probably offer little insight into the biology of these cells).</p>
</div>
<div class="section level2">
<h2 id="visualizing-the-cell-matrix">Visualizing the cell matrix<a class="anchor" aria-label="anchor" href="#visualizing-the-cell-matrix"></a>
</h2>
<!-- Each cell in the single-cell data set is represented as a linear
combination of the $K$ factors. -->
<p>In this matrix factorization, the rows of the <span class="math inline">\({\bf L}\)</span> matrix contains the estimated
“membership levels,” or “memberships,” for each cell. For example:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cell_ids</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"GATATATGTCAGTG-1-b_cells"</span>,</span>
<span>              <span class="st">"GACAGTACCTGTGA-1-memory_t"</span>,</span>
<span>              <span class="st">"TGAAGCACACAGCT-1-b_cells"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="va">fit</span><span class="op">$</span><span class="va">L_pm</span><span class="op">[</span><span class="va">cell_ids</span>, <span class="op">]</span>, digits <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span>
<span><span class="co">#&gt;                            [,1] [,2] [,3] [,4]  [,5]  [,6] [,7] [,8]</span></span>
<span><span class="co">#&gt; GATATATGTCAGTG-1-b_cells  0.562    0    0    0 0.001 0.487    0    0</span></span>
<span><span class="co">#&gt; GACAGTACCTGTGA-1-memory_t 0.636    0    0    0 0.247 0.000    0    0</span></span>
<span><span class="co">#&gt; TGAAGCACACAGCT-1-b_cells  0.617    0    0    0 0.000 0.211    0    0</span></span></code></pre></div>
<p>The first and third cells mainly have membership in factors 1 and 6,
whereas the second cell mostly has membership in factors 1 and 5. Most
of the memberships are zero.</p>
<p>We can use plotting functions provided by <code>flashier</code> to
visualize the relationship between the memberships and the cell labels,
and possibly get some clues about the biological meaning of the factors.
The cell labels are the “sorted” cell subpopulations:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">pbmc_facs</span><span class="op">$</span><span class="va">samples</span><span class="op">$</span><span class="va">subpop</span><span class="op">)</span></span>
<span><span class="co">#&gt;  B cell   CD14+   CD34+ NK cell  T cell </span></span>
<span><span class="co">#&gt;     767     163     687     673    1484</span></span></code></pre></div>
<p>We can produce different types of plots by specifying the
<code>plot_type</code> argument to the <code>plot</code> method. For
example, plotting overlapping histograms makes it immediately clear that
each component 2 through 6 is primarily capturing a single cell
type.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">fit</span>, </span>
<span>     plot_type <span class="op">=</span> <span class="st">"histogram"</span>,</span>
<span>     pm_which <span class="op">=</span> <span class="st">"loadings"</span>, </span>
<span>     pm_groups <span class="op">=</span> <span class="va">pbmc_facs</span><span class="op">$</span><span class="va">samples</span><span class="op">$</span><span class="va">subpop</span>,</span>
<span>     bins <span class="op">=</span> <span class="fl">20</span><span class="op">)</span></span></code></pre></div>
<p><img src="flashier_single_cell_files/figure-html/histograms-1.png" width="600" style="display: block; margin: auto;"></p>
<p>Other plot types can capture more subtle structure. An especially
useful plot type for visualizing nonnegative factorizations is the
“structure plot” — a stacked bar plot in which each component is
represented as a bar of a different color, and the bar heights are given
by the loadings on each component.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">fit</span>, </span>
<span>     plot_type <span class="op">=</span> <span class="st">"structure"</span>,</span>
<span>     pm_which <span class="op">=</span> <span class="st">"loadings"</span>, </span>
<span>     pm_groups <span class="op">=</span> <span class="va">pbmc_facs</span><span class="op">$</span><span class="va">samples</span><span class="op">$</span><span class="va">subpop</span>,</span>
<span>     gap <span class="op">=</span> <span class="fl">25</span><span class="op">)</span></span></code></pre></div>
<p><img src="flashier_single_cell_files/figure-html/structure-plot-1-1.png" width="600" style="display: block; margin: auto;"></p>
<p>Factor 1 (“k1” in the legend) is present in all cell types, and is
likely capturing a “baseline” level of expression throughout. To focus
on differences between cell types, let’s remove this first factor from
our structure plot:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">fit</span>, </span>
<span>     plot_type <span class="op">=</span> <span class="st">"structure"</span>,</span>
<span>     kset <span class="op">=</span> <span class="fl">2</span><span class="op">:</span><span class="fl">8</span>,</span>
<span>     pm_which <span class="op">=</span> <span class="st">"loadings"</span>, </span>
<span>     pm_groups <span class="op">=</span> <span class="va">pbmc_facs</span><span class="op">$</span><span class="va">samples</span><span class="op">$</span><span class="va">subpop</span>,</span>
<span>     gap <span class="op">=</span> <span class="fl">25</span><span class="op">)</span></span></code></pre></div>
<p><img src="flashier_single_cell_files/figure-html/structure-plot-2-1.png" width="600" style="display: block; margin: auto;"></p>
<p>It is clear from this plot that factors 2 through 6 (“k2” through
“k6” in the legend) are capturing, respectively, natural killer (NK)
cells, CD14+ cells, CD34+ cells, T cells and B cells. There is also more
subtle structure captured by the memberships, such as the subset of T
cells with mixed memberships (factors 2 and 5) which suggests a
subpopulation that is intermediate between “pure” T and “pure” NK cells.
It is also interesting to note the subset of CD34+ cells that are
primarily loaded on factor 3; some of these cells may be mislabeled.</p>
<p>We can also visualize the cell matrix in a heatmap:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">fit</span>, </span>
<span>     plot_type <span class="op">=</span> <span class="st">"heatmap"</span>,</span>
<span>     pm_which <span class="op">=</span> <span class="st">"loadings"</span>, </span>
<span>     pm_groups <span class="op">=</span> <span class="va">pbmc_facs</span><span class="op">$</span><span class="va">samples</span><span class="op">$</span><span class="va">subpop</span>,</span>
<span>     gap <span class="op">=</span> <span class="fl">25</span><span class="op">)</span></span></code></pre></div>
<p><img src="flashier_single_cell_files/figure-html/heatmap-1.png" width="390" style="display: block; margin: auto;"></p>
</div>
<div class="section level2">
<h2 id="visualizing-the-gene-matrix">Visualizing the gene matrix<a class="anchor" aria-label="anchor" href="#visualizing-the-gene-matrix"></a>
</h2>
<p>Above, we used the provided cell labels to interpret five of the
factors as capturing distinct cell types (B cells, T cells, etc).
However, one might not always have cell labels, or the existing cell
labels may not be informative for the factors. We can also look to the
<span class="math inline">\({\bf F}\)</span> matrix—the “gene
matrix”—for interpreting the factors.</p>
<p>The gene matrix contains estimates of gene expression changes.
Because these estimates are based on the shifted log counts, more
precisely they are <em>changes in the shifted log-expression.</em> The
first factor captures a “baseline” level of expression, so, broadly
speaking, we interpret these changes relative to this baseline.</p>
<p>Genes that were estimated to have the largest increases in expression
might give the most helpful clues about the underlying biology. For
example, many of the genes with the largest expression increases in
factor 6 are also genes characteristic of B cells:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ldf.html">ldf</a></span><span class="op">(</span><span class="va">fit</span>, type <span class="op">=</span> <span class="st">"i"</span><span class="op">)</span></span>
<span><span class="cn">F</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/with.html" class="external-link">with</a></span><span class="op">(</span><span class="va">res</span>, <span class="cn">F</span> <span class="op"><a href="https://rdrr.io/r/base/matmult.html" class="external-link">%*%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/diag.html" class="external-link">diag</a></span><span class="op">(</span><span class="va">D</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="cn">F</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">pbmc_facs</span><span class="op">$</span><span class="va">genes</span><span class="op">$</span><span class="va">symbol</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="cn">F</span><span class="op">[</span>,<span class="fl">6</span><span class="op">]</span>, decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>, n <span class="op">=</span> <span class="fl">16</span><span class="op">)</span></span>
<span><span class="co">#&gt;      CD74   HLA-DRA  HLA-DRB1     CD79A  HLA-DPB1  HLA-DPA1     CD79B  HLA-DRB5 </span></span>
<span><span class="co">#&gt; 3.4724286 2.9144905 2.3949018 2.2259586 2.1084118 1.9673726 1.8776798 1.6999751 </span></span>
<span><span class="co">#&gt;      CD37       LTB  HLA-DQA1  HLA-DQA2     MS4A1      CD52  HLA-DQB1     TCL1A </span></span>
<span><span class="co">#&gt; 1.6019631 1.4980763 1.3626036 1.1202287 1.1135809 1.0048497 0.9848164 0.9128516</span></span></code></pre></div>
<p>Let’s now apply this simple rule of thumb and examine the top 4 genes
(by expression increase) for factors 2 through 6:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">top_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="cn">F</span>, <span class="fl">2</span>, <span class="va">order</span>, decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">4</span>, <span class="fl">2</span><span class="op">:</span><span class="fl">6</span><span class="op">]</span></span>
<span><span class="va">top_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">fit</span><span class="op">$</span><span class="va">F_pm</span><span class="op">)</span><span class="op">[</span><span class="va">top_genes</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">fit</span>,</span>
<span>     plot_type <span class="op">=</span> <span class="st">"heatmap"</span>,</span>
<span>     pm_which <span class="op">=</span> <span class="st">"factors"</span>,</span>
<span>     pm_subset <span class="op">=</span> <span class="va">top_genes</span>,</span>
<span>     pm_groups <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">top_genes</span>, levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rev.html" class="external-link">rev</a></span><span class="op">(</span><span class="va">top_genes</span><span class="op">)</span><span class="op">)</span>,</span>
<span>     kset <span class="op">=</span> <span class="fl">2</span><span class="op">:</span><span class="fl">6</span>,</span>
<span>     gap <span class="op">=</span> <span class="fl">0.2</span><span class="op">)</span></span></code></pre></div>
<p><img src="flashier_single_cell_files/figure-html/top-genes-all-1.png" width="300" style="display: block; margin: auto;"></p>
<p>Indeed, this approach reveals genes characteristic of the other cell
types (e.g., <em>GNLY</em> for NK cells, <em>S100A9</em> for CD14+
cells), although not always; for example, some genes show large
expression increases more than one factor.</p>
<p>While selecting genes with the largest expression increases appears
to work well here, in other settings there may be more effective
approaches.</p>
<p>Plot of mean vs. change for factor 6 (“B cells”), with top 10 genes
(by largest increase):</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">fit</span>,</span>
<span>     plot_type <span class="op">=</span> <span class="st">"scatter"</span>,</span>
<span>     pm_which <span class="op">=</span> <span class="st">"factors"</span>,</span>
<span>     kset <span class="op">=</span> <span class="fl">6</span>,</span>
<span>     labels <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>     n_labels <span class="op">=</span> <span class="fl">10</span>,</span>
<span>     label_size <span class="op">=</span> <span class="fl">2.5</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"increase in shifted log expression"</span>,</span>
<span>       y <span class="op">=</span> <span class="st">"mean shifted log expression"</span><span class="op">)</span> </span></code></pre></div>
<p><img src="flashier_single_cell_files/figure-html/plot-change-vs-mean-1.png" width="360" style="display: block; margin: auto;"></p>
</div>
<div class="section level2">
<h2 id="other-notes">Other notes<a class="anchor" aria-label="anchor" href="#other-notes"></a>
</h2>
<p>While we have focussed on NMF, flashier is very flexible, and other
types of matrix factorizations are possible. One alternative to NMF that
could potentially reveal other interesting substructures is a
<em>semi-non-negative matrix factorization</em> (semi-NMF). This is
achieved by assigning different priors to <span class="math inline">\({\bf L}\)</span> and <span class="math inline">\({\bf F}\)</span>: a prior with non-negative
support (such as the point-exponential prior) for <span class="math inline">\({\bf L}\)</span>; and a prior with support for all
real numbers for <span class="math inline">\({\bf F}\)</span> (such as
the point-Laplace prior). The call to the “flash” function looks the
same as above except for the “ebnm_fn” argument:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fit_snmf</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/flash.html">flash</a></span><span class="op">(</span><span class="va">shifted_log_counts</span>, </span>
<span>                  ebnm_fn <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">ebnm_point_exponential</span>, <span class="va">ebnm_point_laplace</span><span class="op">)</span>,</span>
<span>                  var_type <span class="op">=</span> <span class="fl">2</span>, </span>
<span>                  greedy_Kmax <span class="op">=</span> <span class="fl">8</span>, </span>
<span>                  S <span class="op">=</span> <span class="va">s1</span>,</span>
<span>                  backfit <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<p>This is the version of R and the packages that were used to generate
these results.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.3.3 (2024-02-29)</span></span>
<span><span class="co">#&gt; Platform: aarch64-apple-darwin20 (64-bit)</span></span>
<span><span class="co">#&gt; Running under: macOS Sonoma 14.5</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRblas.0.dylib </span></span>
<span><span class="co">#&gt; LAPACK: /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.11.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt; [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: America/Chicago</span></span>
<span><span class="co">#&gt; tzcode source: internal</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt; [1] ggplot2_3.5.0      fastTopics_0.6-184 Matrix_1.6-5       flashier_1.0.53   </span></span>
<span><span class="co">#&gt; [5] ebnm_1.1-34       </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;  [1] tidyselect_1.2.1     viridisLite_0.4.2    dplyr_1.1.4         </span></span>
<span><span class="co">#&gt;  [4] farver_2.1.1         fastmap_1.1.1        lazyeval_0.2.2      </span></span>
<span><span class="co">#&gt;  [7] digest_0.6.34        lifecycle_1.0.4      invgamma_1.1        </span></span>
<span><span class="co">#&gt; [10] magrittr_2.0.3       compiler_4.3.3       rlang_1.1.3         </span></span>
<span><span class="co">#&gt; [13] sass_0.4.8           progress_1.2.3       tools_4.3.3         </span></span>
<span><span class="co">#&gt; [16] utf8_1.2.4           yaml_2.3.8           data.table_1.15.2   </span></span>
<span><span class="co">#&gt; [19] knitr_1.45           prettyunits_1.2.0    labeling_0.4.3      </span></span>
<span><span class="co">#&gt; [22] htmlwidgets_1.6.4    scatterplot3d_0.3-44 RColorBrewer_1.1-3  </span></span>
<span><span class="co">#&gt; [25] Rtsne_0.17           withr_3.0.0          purrr_1.0.2         </span></span>
<span><span class="co">#&gt; [28] desc_1.4.3           grid_4.3.3           fansi_1.0.6         </span></span>
<span><span class="co">#&gt; [31] colorspace_2.1-0     scales_1.3.0         gtools_3.9.5        </span></span>
<span><span class="co">#&gt; [34] cli_3.6.2            rmarkdown_2.26       crayon_1.5.2        </span></span>
<span><span class="co">#&gt; [37] ragg_1.2.7           generics_0.1.3       RcppParallel_5.1.7  </span></span>
<span><span class="co">#&gt; [40] httr_1.4.7           pbapply_1.7-2        cachem_1.0.8        </span></span>
<span><span class="co">#&gt; [43] splines_4.3.3        parallel_4.3.3       softImpute_1.4-1    </span></span>
<span><span class="co">#&gt; [46] vctrs_0.6.5          jsonlite_1.8.8       hms_1.1.3           </span></span>
<span><span class="co">#&gt; [49] mixsqp_0.3-54        ggrepel_0.9.5        irlba_2.3.5.1       </span></span>
<span><span class="co">#&gt; [52] horseshoe_0.2.0      systemfonts_1.0.6    trust_0.1-8         </span></span>
<span><span class="co">#&gt; [55] plotly_4.10.4        jquerylib_0.1.4      tidyr_1.3.1         </span></span>
<span><span class="co">#&gt; [58] glue_1.7.0           pkgdown_2.0.7        cowplot_1.1.3       </span></span>
<span><span class="co">#&gt; [61] uwot_0.1.16          Polychrome_1.5.1     gtable_0.3.4        </span></span>
<span><span class="co">#&gt; [64] quadprog_1.5-8       munsell_0.5.0        tibble_3.2.1        </span></span>
<span><span class="co">#&gt; [67] pillar_1.9.0         htmltools_0.5.7      truncnorm_1.0-9     </span></span>
<span><span class="co">#&gt; [70] R6_2.5.1             textshaping_0.3.7    evaluate_0.23       </span></span>
<span><span class="co">#&gt; [73] lattice_0.22-5       highr_0.10           RhpcBLASctl_0.23-42 </span></span>
<span><span class="co">#&gt; [76] memoise_2.0.1        SQUAREM_2021.1       ashr_2.2-66         </span></span>
<span><span class="co">#&gt; [79] bslib_0.6.1          Rcpp_1.0.12          deconvolveR_1.2-1   </span></span>
<span><span class="co">#&gt; [82] xfun_0.42            fs_1.6.3             pkgconfig_2.0.3</span></span></code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Jason Willwerscheid, Peter Carbonetto, Wei Wang, Matthew Stephens.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
