<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>A closer look at flashier • flashier</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="A closer look at flashier">
<meta property="og:description" content="flashier">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">flashier</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.0.53</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">Home</a>
</li>
<li>
  <a href="../articles/index.html">Vignettes</a>
</li>
<li>
  <a href="../reference/index.html">Functions</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/willwerscheid/flashier" class="external-link">Source</a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>A closer look at flashier</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/willwerscheid/flashier/blob/HEAD/vignettes/flashier_closer_look.Rmd" class="external-link"><code>vignettes/flashier_closer_look.Rmd</code></a></small>
      <div class="hidden name"><code>flashier_closer_look.Rmd</code></div>

    </div>

    
    
<p>In addition to <code>flashier</code> and <code>ggplot2</code>, we
will make use of package <code>cowplot</code> below in order to arrange
plots into grids:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/willwerscheid/flashier" class="external-link">flashier</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; Loading required package: ebnm</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://wilkelab.org/cowplot/" class="external-link">cowplot</a></span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="the-ebmf-model">The EBMF Model<a class="anchor" aria-label="anchor" href="#the-ebmf-model"></a>
</h2>
<p>Empirical Bayes matrix factorization (Wang and Stephens, 2021)
decomposes a data matrix <span class="math inline">\(Y \in \mathbb{R}^{n
\times p}\)</span> as <span class="math display">\[ Y = LF' + E,
\]</span> with “loadings” <span class="math inline">\(L \in
\mathbb{R}^{n \times K}\)</span>, “factors” <span class="math inline">\(F \in \mathbb{R}^{p \times K}\)</span>, and
residual errors <span class="math display">\[ e_{ij} \sim N(0, 1 /
\tau_{ij}).\]</span> The model puts priors on each factor and each set
of loadings: <span class="math display">\[f_{\cdot k} \sim g^{(f)}_k,\
\ell_{\cdot k} \sim g^{(\ell)}_k,\]</span> with <span class="math inline">\(g^{(\ell)}_k\)</span> and <span class="math inline">\(g^{(f)}_k\)</span> assumed to belong to some
families of distributions <span class="math inline">\(\mathcal{G}^{(\ell)}\)</span> and <span class="math inline">\(\mathcal{G}^{(f)}\)</span> and estimated using the
data. The default choice of prior family for both factors and loadings
is the family of point-normal distributions: <span class="math display">\[ g \sim \pi_0 \delta_0 + (1 - \pi_0) N(0,
\sigma^2), \]</span> where both <span class="math inline">\(\pi_0\)</span> and <span class="math inline">\(\sigma^2\)</span> are free parameters. This family
is especially useful when factors and loadings can be expected to be at
least modestly sparse; in other settings, different prior families might
be preferred.</p>
<p>To avoid over-parametrization, it is necessary to make some
assumptions about the precision parameters <span class="math inline">\(\tau_{ij}\)</span>. The default assumption is that
all <span class="math inline">\(\tau_{ij}\)</span>s are equal: <span class="math display">\[ e_{ij} \sim N(0, 1 / \tau).\]</span></p>
<p>Note that when the prior families <span class="math inline">\(\mathcal{G}^{(\ell)}\)</span> and <span class="math inline">\(\mathcal{G}^{(f)}\)</span> are closed under
scaling (as is typically the case), then the model as formulated above
is not identifiable, since we can scale each set of loadings <span class="math inline">\(\ell_{\cdot k}\)</span> by any constant <span class="math inline">\(d_k \ne 0\)</span> if we also scale factor <span class="math inline">\(f_{\cdot k}\)</span> by <span class="math inline">\(1 / d_k\)</span>.</p>
<p>We can make the matrix factorization unique by writing <span class="math display">\[ Y = LDF' + E, \]</span> with the scales of
loadings <span class="math inline">\(\ell_{\cdot 1}, \ldots, \ell_{\cdot
K}\)</span> and factors <span class="math inline">\(f_{\cdot 1}, \ldots,
f_{\cdot K}\)</span> constrained in some fashion (for example, by
requiring <span class="math inline">\(\| \ell_{\cdot k} \|_2 =
1\)</span> and <span class="math inline">\(\| f_{\cdot k} \|_2 =
1\)</span> for all <span class="math inline">\(k\)</span>). We refer to
such a factorization as an <strong>LDF factorization</strong>.</p>
</div>
<div class="section level2">
<h2 id="the-gtex-dataset-and-the-flash-function">The <code>gtex</code> dataset and the <code>flash()</code>
function<a class="anchor" aria-label="anchor" href="#the-gtex-dataset-and-the-flash-function"></a>
</h2>
<p>Throughout this vignette, we will use the main <code><a href="../reference/flash.html">flash()</a></code>
function to produce a number of EBMF fits to dataset <code>gtex</code>,
which is comprised of a subset of the data used as the primary example
in Wang and Stephens (2021). The dataset is derived from data made
available by the Genotype Tissue Expression (GTEx) project (Lonsdale et
al. 2013), which provides <span class="math inline">\(z\)</span>-scores
for assessing the significance of effects of genetic variants (“single
nucleotide polymorphisms” or SNPs) on gene expression across 44 human
tissues. To reduce the data to a more manageable size, Urbut et
al. (2019) chose the “top” SNP for each gene — that is, the SNP
associated with the largest (absolute) <span class="math inline">\(z\)</span>-score over all 44 tissues. This process
yields a 16,069 by 44 matrix of <span class="math inline">\(z\)</span>-scores, with rows corresponding to
SNP-gene pairs or “eQTLs” and columns corresponding to tissues. The
<code>gtex</code> dataset included in <code>flashier</code> is further
subsampled down to 1000 rows.</p>
<p>One potential aim in applying matrix factorization methods to the
<code>gtex</code> dataset is to analyze patterns of effect sharing. For
example, Wang and Stephens (2021) point out an EBMF factor that has
large values across brain tissues (cortex, cerebellum, basal ganglia,
etc.), but values much closer to zero for all non-brain tissues. We can
expect eQTLs that are loaded on this factor to affect brain and
non-brain tissues much differently. Other factors have large values for
a single tissue only (e.g., testis). We can expect eQTLs loaded on such
factors to be more or less tissue-specific.</p>
<p>To fit the EBMF model, we use function <code><a href="../reference/flash.html">flash()</a></code>, which
adds up to <code>greedy_Kmax</code> factor/loadings pairs using the
“greedy” algorithm described in Wang and Stephens (2021). In brief, the
algorithm adds new factor/loadings pairs one at a time, optimizing each
pair as it gets added without returning to re-optimize previously added
pairs. In other words, it finds the best rank-one fit; then the best
rank-two fit with the first factor/loadings pair fixed; then the best
rank-three fit with the first two factor/loadings pairs fixed; and so
on. The algorithm stops adding factor/loadings pairs when they fail to
yield an increase in the variational lower bound on the log likelihood
(or ELBO), and in this sense EBMF can “automatically” select the number
of factor/loadings pairs <span class="math inline">\(K\)</span>. Thus it
is often sufficient simply to set <code>greedy_Kmax</code> as high as is
practically manageable (the default is <code>greedy_Kmax = 50</code>).
Here, however, we will restrict the number of factor/loadings pairs in
order to ensure that the vignette can be run quickly:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">gtex</span>, <span class="va">gtex_colors</span><span class="op">)</span></span>
<span><span class="va">fit_default</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/flash.html">flash</a></span><span class="op">(</span><span class="va">gtex</span>, greedy_Kmax <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span>
<span><span class="co">#&gt; Adding factor 1 to flash object...</span></span>
<span><span class="co">#&gt; Adding factor 2 to flash object...</span></span>
<span><span class="co">#&gt; Adding factor 3 to flash object...</span></span>
<span><span class="co">#&gt; Adding factor 4 to flash object...</span></span>
<span><span class="co">#&gt; Adding factor 5 to flash object...</span></span>
<span><span class="co">#&gt; Wrapping up...</span></span>
<span><span class="co">#&gt; Done.</span></span>
<span><span class="co">#&gt; Nullchecking 5 factors...</span></span>
<span><span class="co">#&gt; Done.</span></span></code></pre></div>
<p>By default, <code><a href="../reference/flash.html">flash()</a></code> performs a “nullcheck” after the
greedy algorithm terminates: that is, it loops over each greedily added
factor/loadings pair and checks that removing it does not produce an
increase in the ELBO (i.e., a better fit). If it does, then the
factor/loadings pair is removed from the final fit.</p>
<p>The output of function <code><a href="../reference/flash.html">flash()</a></code> is an object of class
<code>flash</code>. Useful information about the fit can be accessed as
list elements. For example, the ELBO can be accessed as:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fit_default</span><span class="op">$</span><span class="va">elbo</span></span>
<span><span class="co">#&gt; [1] -83131.69</span></span></code></pre></div>
<p>Additionally, a few class methods have been implemented, including
<code><a href="https://rdrr.io/r/stats/fitted.values.html" class="external-link">fitted()</a></code>, which returns <span class="math inline">\(\mathbb{E} (LF')\)</span>;
<code><a href="https://rdrr.io/r/stats/residuals.html" class="external-link">residuals()</a></code>, which yields the matrix of expected residuals
<span class="math inline">\(Y - \mathbb{E} (LF')\)</span>; and
<code><a href="../reference/ldf.html">ldf()</a></code>, which gives the LDF factorization with the scaling
of loadings and factors determined by argument <code>type</code>.</p>
<p>Finally, <code>flashier</code> provides a convenient
<code><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot()</a></code> method. The type of plot is controlled by the
<code>plot_type</code> argument. The default option is a “scree plot,”
which shows the proportion of variance explained by each factor/loadings
pair:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">fit_default</span><span class="op">)</span></span></code></pre></div>
<p><img src="flashier_closer_look_files/figure-html/plot.scree-1.png" width="576"></p>
<p>Other plot types, including bar plots, histograms, scatter plots,
heatmaps, and “structure” plots, can provide a more detailed overview of
posterior means for either factors or loadings (depending on the
argument to <code>pm_which</code>). Here, supply colors for the
individual bars in a bar plot in order to match the figures included in
Wang and Stephens (2021):</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">fit_default</span>, </span>
<span>     pm_which <span class="op">=</span> <span class="st">"factors"</span>, </span>
<span>     pm_colors <span class="op">=</span> <span class="va">gtex_colors</span>,</span>
<span>     plot_type <span class="op">=</span> <span class="st">"bar"</span><span class="op">)</span></span></code></pre></div>
<p><img src="flashier_closer_look_files/figure-html/plot.bar-1.png" width="576"></p>
<p>All of the plot types are available via a dual interface. In addition
to being able to specify the <code>plot_type</code> argument in the
<code>plot</code> method, each plot type has its own
<code>flash_plot_xxx</code> function, which typically exposes a few more
arguments. Further, since a <code>ggplot2</code> object is returned by
all plotting functions, additional customization is straightforward:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/flash_plot_bar.html">flash_plot_bar</a></span><span class="op">(</span><span class="va">fit_default</span>, </span>
<span>               pm_which <span class="op">=</span> <span class="st">"factors"</span>, </span>
<span>               pm_colors <span class="op">=</span> <span class="va">gtex_colors</span>,</span>
<span>               labels <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>               ncol <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">6</span>, angle <span class="op">=</span> <span class="fl">60</span>, color <span class="op">=</span> <span class="va">gtex_colors</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="flashier_closer_look_files/figure-html/plot.bar.custom-1.png" width="576"></p>
<p>Most of the large values for factor 2 above correspond to brain
tissues (yellow), but the factor is not very “clean” in the sense that
many other tissues have medium to large values as well. Factor 3 is
primarily confined to whole blood, but there again appears to be a lot
of noise, which makes biological interpretation difficult.</p>
<p>In the following sections, we will attempt to clean up these factors
by choosing better (non-default) arguments to various parameters in
function <code><a href="../reference/flash.html">flash()</a></code>. We note in passing that
<code>flashier</code> supplies two interfaces for fitting EBMF models:
one, via <code><a href="../reference/flash.html">flash()</a></code>; and a second, pipeable interface that
provides many more options for customization. In this vignette, we only
consider the first, simpler interface; for an introduction to the
second, see the “Advanced flashier” vignette.</p>
</div>
<div class="section level2">
<h2 id="variance-structures">Variance structures<a class="anchor" aria-label="anchor" href="#variance-structures"></a>
</h2>
<p>We begin by considering parameter <code>var_type</code>. As mentioned
above, the default model (<code>var_type = 0</code>) estimates a single
precision parameter <span class="math inline">\(\tau\)</span>: <span class="math display">\[ e_{ij} \sim N(0, 1 / \tau).\]</span> It is also
possible to estimate a precision parameter for each row
(<code>var_type = 1</code>) or for each column
(<code>var_type = 2</code>). For example, setting
<code>var_type = 2</code> puts <span class="math display">\[ e_{ij} \sim
N(0, 1 / \tau_j),\]</span> so that residual variance is assumed to vary
from column to column but is constant within each column. Note that this
“by-column” assumption contains the constant variance assumption
<code>var_type = 0</code> as a special case, so it will typically yield
a higher ELBO (we write “typically” rather than “always” because the
optimization surface is in general non-convex, so there is no guarantee
that the global optimum will be found):</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fit_var_bycol</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/flash.html">flash</a></span><span class="op">(</span><span class="va">gtex</span>, greedy_Kmax <span class="op">=</span> <span class="fl">5</span>, var_type <span class="op">=</span> <span class="fl">2</span>, verbose <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>const_elbo <span class="op">=</span> <span class="va">fit_default</span><span class="op">$</span><span class="va">elbo</span>, bycol_elbo <span class="op">=</span> <span class="va">fit_var_bycol</span><span class="op">$</span><span class="va">elbo</span><span class="op">)</span></span>
<span><span class="co">#&gt; const_elbo bycol_elbo </span></span>
<span><span class="co">#&gt;  -83131.69  -80490.27</span></span></code></pre></div>
<p>The most general possible assumption allowed by <code>flashier</code>
is that the matrix of precision parameters <span class="math inline">\(\tau_{ij}\)</span> is an arbitrary rank-one
matrix. That is, for each <span class="math inline">\(i\)</span> and
<span class="math inline">\(j\)</span>, <span class="math inline">\(\tau_{ij} = \tau^{(1)}_i \tau^{(2)}_j\)</span>,
where <span class="math inline">\(\tau^{(1)}\)</span> is a <span class="math inline">\(n\)</span>-vector to be estimated and <span class="math inline">\(\tau^{(2)}\)</span> is a <span class="math inline">\(p\)</span>-vector to be estimated. This
assumption, which we refer to as the “Kronecker” variance assumption,
can be motivated by noting that it describes a model in which residuals
are distributed <span class="math inline">\(e_{ij} \sim N(0,
1)\)</span>, but the rows and columns of <span class="math inline">\(Y\)</span> have each been scaled by some (unknown)
constants: <span class="math display">\[ \text{diag} (\tau^{(1)}) Y
\text{diag} (\tau^{(2)}) = L F' + E.\]</span></p>
<p>Just as the by-column assumption usually yields an increase in ELBO
relative to the constant variance assumption, the Kronecker assumption
will generally yield a higher ELBO than the by-column assumption:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fit_var_kronecker</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/flash.html">flash</a></span><span class="op">(</span><span class="va">gtex</span>, greedy_Kmax <span class="op">=</span> <span class="fl">5</span>, var_type <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span>, verbose <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>bycol_elbo <span class="op">=</span> <span class="va">fit_var_bycol</span><span class="op">$</span><span class="va">elbo</span>, kron_elbo <span class="op">=</span> <span class="va">fit_var_kronecker</span><span class="op">$</span><span class="va">elbo</span><span class="op">)</span></span>
<span><span class="co">#&gt; bycol_elbo  kron_elbo </span></span>
<span><span class="co">#&gt;  -80490.27  -78636.44</span></span></code></pre></div>
<p>Although the Kronecker assumption can yield large increases in ELBO,
one should be careful using it with very large datasets, as the
precision parameters must be estimated at each iteration via an
alternating maximization algorithm. In contrast, precision parameters
can be obtained analytically (and thus much more quickly) when using a
constant, by-row, or by-column variance assumption. Note also that
despite the large increase in ELBO from the by-column to the Kronecker
variance assumption, there are not obvious differences in the plotted
posterior means:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span></span>
<span>  <span class="va">fit_var_bycol</span>, </span>
<span>  pm_which <span class="op">=</span> <span class="st">"factors"</span>, </span>
<span>  pm_colors <span class="op">=</span> <span class="va">gtex_colors</span>, </span>
<span>  plot_type <span class="op">=</span> <span class="st">"bar"</span></span>
<span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"By-column variance assumption"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span></span>
<span>  <span class="va">fit_var_kronecker</span>, </span>
<span>  pm_which <span class="op">=</span> <span class="st">"factors"</span>, </span>
<span>  pm_colors <span class="op">=</span> <span class="va">gtex_colors</span>, </span>
<span>  plot_type <span class="op">=</span> <span class="st">"bar"</span></span>
<span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Kronecker variance assumption"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html" class="external-link">plot_grid</a></span><span class="op">(</span><span class="va">p1</span>, <span class="va">p2</span>, nrow <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<p><img src="flashier_closer_look_files/figure-html/gtex.bycol.plot-1.png" width="576"></p>
<p>To compare the fits more directly, we can use function
<code><a href="../reference/ldf.html">ldf()</a></code> to extract factor values, with argument
<code>type</code> set to <code>"m"</code> in order to match the values
displayed by <code>plot.flash</code> (<code>"m"</code> stands for
maximum norm; setting <code>type = "m"</code> scales the columns of
<span class="math inline">\(F\)</span> so that the maximum absolute
value over each column is equal to 1). A scatterplot confirms that
estimates are very similar, with minor differences in factors 2, 4, and
5:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">comparison_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>  var_by_col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/vector.html" class="external-link">as.vector</a></span><span class="op">(</span><span class="fu"><a href="../reference/ldf.html">ldf</a></span><span class="op">(</span><span class="va">fit_var_bycol</span>, type <span class="op">=</span> <span class="st">"m"</span><span class="op">)</span><span class="op">$</span><span class="cn">F</span><span class="op">)</span>,</span>
<span>  var_kronecker <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/vector.html" class="external-link">as.vector</a></span><span class="op">(</span><span class="fu"><a href="../reference/ldf.html">ldf</a></span><span class="op">(</span><span class="va">fit_var_kronecker</span>, type <span class="op">=</span> <span class="st">"m"</span><span class="op">)</span><span class="op">$</span><span class="cn">F</span><span class="op">)</span>,</span>
<span>  k <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, each <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">gtex</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">comparison_df</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">var_by_col</span>, y <span class="op">=</span> <span class="va">var_kronecker</span>, col <span class="op">=</span> <span class="va">k</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>alpha <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_abline</a></span><span class="op">(</span>slope <span class="op">=</span> <span class="fl">1</span>, linetype <span class="op">=</span> <span class="st">"dashed"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html" class="external-link">scale_color_brewer</a></span><span class="op">(</span>palette <span class="op">=</span> <span class="st">"Set1"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"By-column variance assumption"</span>, y <span class="op">=</span> <span class="st">"Kronecker variance assumption"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_minimal</a></span><span class="op">(</span><span class="op">)</span> </span></code></pre></div>
<p><img src="flashier_closer_look_files/figure-html/var.compare-1.png" width="576"></p>
</div>
<div class="section level2">
<h2 id="measurement-error">Measurement error<a class="anchor" aria-label="anchor" href="#measurement-error"></a>
</h2>
<p>In many scenarios, the data <span class="math inline">\(Y\)</span> is
observed with some known error. In such a case, it might be preferable
to fit the model</p>
<p><span class="math display">\[ Y = L F' + S + E, \]</span> where
<span class="math inline">\(S_{ij} \sim N(0, s^2_{ij})\)</span> with the
<span class="math inline">\(s^2_{ij}\)</span>s fixed.
<!--In other words, one might prefer to fit the model
$$ Y_{ij} \sim N \left(\sum_k \ell_{ik} f_{jk}, s^2_{ij} + 1 / \tau_{ij}\right). $$
--></p>
<p>In simple cases, this model can <em>almost</em> be reduced to the
model described above. For example, since the <code>gtex</code> dataset
is comprised of <span class="math inline">\(z\)</span>-scores, we might
choose to set the <span class="math inline">\(s_{ij}\)</span>s
identically equal to one. With by-column precision parameters <span class="math inline">\(\tau_j\)</span>, this yields the model <span class="math display">\[ Y_{ij} \sim N \left(\sum_k \ell_{ik} f_{jk}, 1 /
\tau^2_{j} + 1 \right). \]</span> This is equivalent to the previously
described model <span class="math display">\[ Y_{ij} \sim N \left(\sum_k
\ell_{ik} f_{jk}, 1 / \tilde{\tau}^2_{j} \right) \]</span> provided that
we add constraints <span class="math inline">\(\tilde{\tau_j} \le
1\)</span> for all <span class="math inline">\(j\)</span>. This model
should yield a lower ELBO than the model with arbitrary column-specific
variances (again, we write “should” because it is not guaranteed that
<code><a href="../reference/flash.html">flash()</a></code> will converge to a global optimum), but it is
arguably the more correct model.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fit_with_meas_err</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/flash.html">flash</a></span><span class="op">(</span><span class="va">gtex</span>, S <span class="op">=</span> <span class="fl">1</span>, greedy_Kmax <span class="op">=</span> <span class="fl">5</span>, var_type <span class="op">=</span> <span class="fl">2</span>, verbose <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>bycol_elbo <span class="op">=</span> <span class="va">fit_var_bycol</span><span class="op">$</span><span class="va">elbo</span>, meas_err_elbo <span class="op">=</span> <span class="va">fit_with_meas_err</span><span class="op">$</span><span class="va">elbo</span><span class="op">)</span></span>
<span><span class="co">#&gt;    bycol_elbo meas_err_elbo </span></span>
<span><span class="co">#&gt;     -80490.27     -80491.84</span></span></code></pre></div>
<p>We note here that it is also possible to pass a matrix of standard
errors <span class="math inline">\(s_{ij}\)</span> as argument to
<code>S</code>. If these standard errors are to account for all residual
variance (that is, if the matrix of “true” means can assumed to be
exactly low-rank rather than approximately so), then one can set
<code>var_type = NULL</code>. If not, then updates to precision
parameters <span class="math inline">\(\tau_{ij}\)</span> must be
accomplished via one call to function <code><a href="https://rdrr.io/r/stats/optimize.html" class="external-link">optimize()</a></code> for each
precision parameter, so variance estimation can be very slow. In such
cases, we generally recommend leaving <code>S</code> unspecified and
allowing <em>all</em> residual variance to be estimated.</p>
</div>
<div class="section level2">
<h2 id="prior-families">Prior families<a class="anchor" aria-label="anchor" href="#prior-families"></a>
</h2>
<p>We next turn to parameter <code>ebnm_fn</code>. Note that the
argument to <code>ebnm_fn</code> is a <em>function</em>; indeed, in
<code>flashier,</code> prior families <span class="math inline">\(\mathcal{G}^{(\ell)}_k\)</span> and <span class="math inline">\(\mathcal{G}^{(f)}_k\)</span> are not specified
directly, but rather implicitly via the functions used to solve the
empirical Bayes normal means (EBNM) subproblems, which have form <span class="math display">\[ \begin{gather} x_i \sim \mathcal{N} \left(
\theta_i, s_i^2 \right) \\ \theta_i \sim g \in \mathcal{G}. \end{gather}
\]</span> (Here, the <span class="math inline">\(x_i\)</span>s and <span class="math inline">\(s_i\)</span>s are known observations and standard
errors and the <span class="math inline">\(\theta_i\)</span>s are
unknown “means.”) Effectively, <code>ebnm_fn</code> does not only
(implicitly) specify <span class="math inline">\(\mathcal{G}\)</span>,
but also (directly) the <em>method</em> for estimating <span class="math inline">\(g \in \mathcal{G}\)</span>.</p>
<p>A number of useful EBNM solvers are provided by package
<code>ebnm</code> (Willwerscheid and Stephens 2021). The default
argument is <code>ebnm_fn = ebnm_point_normal</code>, which takes the
prior family <span class="math inline">\(\mathcal{G}\)</span> (for both
factors and loadings) to be the family of point-normal distributions:
<span class="math display">\[ g \sim \pi_0 \delta_0 + (1 - \pi_0) N(0,
\sigma^2). \]</span> In the scenario under consideration, this choice of
prior family makes sense, as it is reasonable to expect some degree of
sparsity in both factors and loadings: some effects might only be shared
across a few related tissues (e.g., brain tissues but not other
tissues), and patterns of sharing may not be ubiquitous across SNP-gene
pairs (e.g., many effects may be constant across brain and non-brain
tissues).</p>
<p>By varying <code>ebnm_fn</code>, we also vary our assumptions about
the “true” underlying distributions of factors and loadings. For
example, if we can reasonably expect each set of loadings <span class="math inline">\(\ell_{\cdot k}\)</span> to be heavy-tailed, then
we might prefer a more flexible family of distributions, such as the
family of scale mixtures of normals
(<code>ebnm_fn = ebnm_normal_scale_mixture</code>) or the family of
symmetric distributions that are unimodal at zero
(<code>ebnm_fn = ebnm_unimodal_symmetric</code>). Note that these
families are nested, so that the unimodal family should yield a higher
ELBO than scale mixtures of normals, which should in turn yield a higher
ELBO than the family of point-normals:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fit_normalmix</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/flash.html">flash</a></span><span class="op">(</span></span>
<span>  <span class="va">gtex</span>, </span>
<span>  greedy_Kmax <span class="op">=</span> <span class="fl">5</span>, </span>
<span>  ebnm_fn <span class="op">=</span> <span class="va">ebnm_normal_scale_mixture</span>, </span>
<span>  verbose <span class="op">=</span> <span class="fl">0</span></span>
<span><span class="op">)</span></span>
<span><span class="va">fit_unimodal</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/flash.html">flash</a></span><span class="op">(</span></span>
<span>  <span class="va">gtex</span>, </span>
<span>  greedy_Kmax <span class="op">=</span> <span class="fl">5</span>, </span>
<span>  ebnm_fn <span class="op">=</span> <span class="va">ebnm_unimodal_symmetric</span>, </span>
<span>  verbose <span class="op">=</span> <span class="fl">0</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>pn_elbo <span class="op">=</span> <span class="va">fit_default</span><span class="op">$</span><span class="va">elbo</span>, </span>
<span>  normalmix_elbo <span class="op">=</span> <span class="va">fit_normalmix</span><span class="op">$</span><span class="va">elbo</span>, </span>
<span>  unimodal_elbo <span class="op">=</span> <span class="va">fit_unimodal</span><span class="op">$</span><span class="va">elbo</span><span class="op">)</span></span>
<span><span class="co">#&gt;        pn_elbo normalmix_elbo  unimodal_elbo </span></span>
<span><span class="co">#&gt;      -83131.69      -82981.84      -82978.90</span></span></code></pre></div>
<p>It is also possible to introduce sign constraints via the choice of
prior family; for example, to constrain factors to be nonnegative, we
need only choose a prior family with nonnegative support such as the
family of point-exponential distributions
(<code>ebnm_fn = ebnm_point_exponential</code>) or the family of all
nonnegative distributions that are unimodal at zero
(<code>ebnm_fn = ebnm_unimodal_nonnegative</code>). This constraint can
model the (reasonable) assumption that when effects are shared across
tissues, they are shared in the same direction (that is, expression
either increases or decreases across a number of related tissues, but
does not increase in some and simultaneously decrease in others). We can
constrain factors while allowing loadings to remain unconstrained by
passing a list as argument to <code>ebnm_fn</code>:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fit_seminonnegative</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/flash.html">flash</a></span><span class="op">(</span></span>
<span>  <span class="va">gtex</span>,</span>
<span>  greedy_Kmax <span class="op">=</span> <span class="fl">5</span>, </span>
<span>  var_type <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  ebnm_fn <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">ebnm_point_normal</span>, <span class="va">ebnm_point_exponential</span><span class="op">)</span>, </span>
<span>  verbose <span class="op">=</span> <span class="fl">0</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>The result is a “semi-nonnegative” factorization, which in the
scenario under consideration yields sparse, interpretable factors
(especially in comparison to the factors produced using all default
arguments to <code><a href="../reference/flash.html">flash()</a></code>; see above). Here, factor 2 represents
sharing among brain, pituitary, and testis tissues; factor 3, among
whole blood, spleen tissue, lung tissue, and lymphocytes; and factor 5,
among heart and muscle tissues. Factor 4 represents effects uniquely
expressed in testis tissue:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">fit_seminonnegative</span>,</span>
<span>     pm_which <span class="op">=</span> <span class="st">"factors"</span>, </span>
<span>     pm_colors <span class="op">=</span> <span class="va">gtex_colors</span>,</span>
<span>     labels <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>     ncol <span class="op">=</span> <span class="fl">1</span>,</span>
<span>     plot_type <span class="op">=</span> <span class="st">"bar"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">6</span>, angle <span class="op">=</span> <span class="fl">60</span>, color <span class="op">=</span> <span class="va">gtex_colors</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="flashier_closer_look_files/figure-html/unnamed-chunk-1-1.png" width="576"></p>
<!-- TODO move to advanced -->
<!-- To pass non-default arguments to functions in package `ebnm`, `flashier` provides the convenience function `flash_ebnm`, which simply passes its arguments to `ebnm`. For example, to fit a normal prior with both mean and variance to be estimated, -->
<!-- $$ g \sim N(\mu, \sigma^2), $$ -->
<!-- one can set `ebnm_fn = flash_ebnm(prior_family = "normal", mode = "estimate")`. The latter can be especially useful for fitting a "mean" factor such as the first factor in the `gtex` example  -->
<!-- (which we wouldn't expect to be centered at zero, since the data is not centered): -->
<!-- ```{r gtex.w.mean} -->
<!-- gtex_intercept <- flash( -->
<!--   gtex,  -->
<!--   greedy_Kmax = 1, -->
<!--   ebnm_fn = flash_ebnm(prior_family = "normal", mode = "estimate"), -->
<!--   verbose = 0 -->
<!-- ) -->
<!-- ``` -->
</div>
<div class="section level2">
<h2 id="backfitting">Backfitting<a class="anchor" aria-label="anchor" href="#backfitting"></a>
</h2>
<p>The “backfitting” algorithm generally requires a large increase in
fitting time but can substantially improve the overall model fit. As
described above, the greedy algorithm leaves previously added
factor/loadings pairs alone when optimizing each newly added
factor/loadings pair. In contrast, the backfitting algorithm iterates
over a number of existing factor/loadings pairs, updating each pair one
at a time until all have converged (see Wang and Stephens, 2021). In
this way, factors and loadings that are added early on can, so to speak,
use information contained in subsequent factors and loadings to improve
the overall model fit. Since, by default, a backfit takes the greedy fit
as its starting point, it is guaranteed to produce an increase in
ELBO:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fit_greedy</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/flash.html">flash</a></span><span class="op">(</span></span>
<span>  <span class="va">gtex</span>, </span>
<span>  greedy_Kmax <span class="op">=</span> <span class="fl">5</span>, </span>
<span>  var_type <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  ebnm_fn <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">ebnm_point_normal</span>, <span class="va">ebnm_point_exponential</span><span class="op">)</span>, </span>
<span>  backfit <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>  verbose <span class="op">=</span> <span class="fl">0</span></span>
<span><span class="op">)</span></span>
<span><span class="va">fit_backfit</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/flash.html">flash</a></span><span class="op">(</span></span>
<span>  <span class="va">gtex</span>, </span>
<span>  greedy_Kmax <span class="op">=</span> <span class="fl">5</span>, </span>
<span>  var_type <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  ebnm_fn <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">ebnm_point_normal</span>, <span class="va">ebnm_point_exponential</span><span class="op">)</span>, </span>
<span>  backfit <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>  verbose <span class="op">=</span> <span class="fl">0</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>greedy_elbo <span class="op">=</span> <span class="va">fit_greedy</span><span class="op">$</span><span class="va">elbo</span>, bf_elbo <span class="op">=</span> <span class="va">fit_backfit</span><span class="op">$</span><span class="va">elbo</span><span class="op">)</span></span>
<span><span class="co">#&gt; greedy_elbo     bf_elbo </span></span>
<span><span class="co">#&gt;   -81194.23   -80893.79</span></span></code></pre></div>
<p>Qualitatively, results can vary: for large, complex fits, the
improvement can be considerable; for smaller fits, there might be few
obvious differences. Interestingly, backfitting the semi-nonnegative fit
from above produces a <em>less</em> visually appealing fit. We suspect
that this occurs because 5 factor/loadings pairs is not enough to
account for the structure in the data; if one increases
<code>greedy_Kmax</code>, one obtains a more satisfying backfit.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span></span>
<span>  <span class="va">fit_greedy</span>, </span>
<span>  pm_which <span class="op">=</span> <span class="st">"factors"</span>,</span>
<span>  pm_colors <span class="op">=</span> <span class="va">gtex_colors</span>, </span>
<span>  plot_type <span class="op">=</span> <span class="st">"bar"</span></span>
<span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Greedy fit"</span><span class="op">)</span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span></span>
<span>  <span class="va">fit_backfit</span>, </span>
<span>  pm_which <span class="op">=</span> <span class="st">"factors"</span>,</span>
<span>  pm_colors <span class="op">=</span> <span class="va">gtex_colors</span>, </span>
<span>  plot_type <span class="op">=</span> <span class="st">"bar"</span></span>
<span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Backfit"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html" class="external-link">plot_grid</a></span><span class="op">(</span><span class="va">p1</span>, <span class="va">p2</span>, nrow <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<p><img src="flashier_closer_look_files/figure-html/backfit.plot-1.png" width="576"></p>
<p>We again obtain a more direct comparison of fits using function
<code><a href="../reference/ldf.html">ldf()</a></code>. Here it is apparent that backfitting makes factor 2
less sparse by “borrowing” from factor 1. Other factors remain very
similar:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">comparison_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>  fit_greedy <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/vector.html" class="external-link">as.vector</a></span><span class="op">(</span><span class="fu"><a href="../reference/ldf.html">ldf</a></span><span class="op">(</span><span class="va">fit_greedy</span>, type <span class="op">=</span> <span class="st">"m"</span><span class="op">)</span><span class="op">$</span><span class="cn">F</span><span class="op">)</span>,</span>
<span>  fit_backfit <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/vector.html" class="external-link">as.vector</a></span><span class="op">(</span><span class="fu"><a href="../reference/ldf.html">ldf</a></span><span class="op">(</span><span class="va">fit_backfit</span>, type <span class="op">=</span> <span class="st">"m"</span><span class="op">)</span><span class="op">$</span><span class="cn">F</span><span class="op">)</span>,</span>
<span>  k <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, each <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">gtex</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">comparison_df</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">fit_greedy</span>, y <span class="op">=</span> <span class="va">fit_backfit</span>, col <span class="op">=</span> <span class="va">k</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>alpha <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_abline</a></span><span class="op">(</span>slope <span class="op">=</span> <span class="fl">1</span>, linetype <span class="op">=</span> <span class="st">"dashed"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html" class="external-link">scale_color_brewer</a></span><span class="op">(</span>palette <span class="op">=</span> <span class="st">"Set1"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"Greedy fit"</span>, y <span class="op">=</span> <span class="st">"Backfit"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_minimal</a></span><span class="op">(</span><span class="op">)</span> </span></code></pre></div>
<p><img src="flashier_closer_look_files/figure-html/bf.compare-1.png" width="576"></p>
</div>
<div class="section level2">
<h2 id="sampling-from-the-posterior">Sampling from the posterior<a class="anchor" aria-label="anchor" href="#sampling-from-the-posterior"></a>
</h2>
<p>One of the list elements in the object returned by <code>flash</code>
is a function that can sample from posterior distributions on loadings
and factors. Take the backfit above as an example. To better understand
which tissues are bound up with whole blood effects, we might like
confidence intervals for the third factor. We construct 95% confidence
intervals using 200 samples and then finding 2.5% and 97.5% quantiles as
follows:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Set seed for reproducibility.</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="co"># Use returned sampler to sample from posterior.</span></span>
<span><span class="va">samp</span> <span class="op">&lt;-</span> <span class="va">fit_backfit</span><span class="op">$</span><span class="fu">sampler</span><span class="op">(</span>nsamp <span class="op">=</span> <span class="fl">200</span><span class="op">)</span></span>
<span><span class="co"># Only keep factor 3.</span></span>
<span><span class="va">factor3_samp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">samp</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="va">x</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">[</span>, <span class="fl">3</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co"># Normalize the loadings.</span></span>
<span><span class="va">factor3_samp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">factor3_samp</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="va">x</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># Get 95% confidence intervals.</span></span>
<span><span class="va">factor3_ci</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="va">factor3_samp</span>, <span class="fl">1</span>, <span class="va">quantile</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.025</span>, <span class="fl">0.975</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Since the <code>plot</code> method returns a <code>ggplot2</code>
object, it can be customized using <code>ggplot</code> syntax, making
the addition of error bars a simple task:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">fit_backfit</span>, kset <span class="op">=</span> <span class="fl">3</span>, pm_colors <span class="op">=</span> <span class="va">gtex_colors</span>, plot_type <span class="op">=</span> <span class="st">"bar"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_linerange.html" class="external-link">geom_errorbar</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>ymin <span class="op">=</span> <span class="va">factor3_ci</span><span class="op">[</span><span class="fl">1</span>, <span class="op">]</span>, ymax <span class="op">=</span> <span class="va">factor3_ci</span><span class="op">[</span><span class="fl">2</span>, <span class="op">]</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="flashier_closer_look_files/figure-html/final.bf.ci.plot-1.png" width="576"></p>
<p>Confidence intervals for spleen tissue, lung tissue, and lymphocytes
do not contain zero, so we can be reasonably confident that this pattern
of sharing is “real” and not just an artefact of the data. We note,
however, that empirical Bayes methods are known to underestimate
uncertainty in the data (since they do not account for uncertainty in
estimates <span class="math inline">\(g \in \mathcal{G}\)</span>), so we
offer this last conclusion with a large grain of salt.</p>
</div>
<div class="section level2">
<h2 id="session-information">Session information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h2>
<p>The following R version and packages were used to generate this
vignette:</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.3.3 (2024-02-29)</span></span>
<span><span class="co">#&gt; Platform: aarch64-apple-darwin20 (64-bit)</span></span>
<span><span class="co">#&gt; Running under: macOS Sonoma 14.5</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRblas.0.dylib </span></span>
<span><span class="co">#&gt; LAPACK: /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.11.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt; [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: America/Chicago</span></span>
<span><span class="co">#&gt; tzcode source: internal</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt; [1] cowplot_1.1.3   ggplot2_3.5.0   flashier_1.0.53 ebnm_1.1-34    </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;  [1] tidyselect_1.2.1     viridisLite_0.4.2    dplyr_1.1.4         </span></span>
<span><span class="co">#&gt;  [4] farver_2.1.1         fastmap_1.1.1        lazyeval_0.2.2      </span></span>
<span><span class="co">#&gt;  [7] digest_0.6.34        lifecycle_1.0.4      invgamma_1.1        </span></span>
<span><span class="co">#&gt; [10] magrittr_2.0.3       compiler_4.3.3       rlang_1.1.3         </span></span>
<span><span class="co">#&gt; [13] sass_0.4.8           progress_1.2.3       tools_4.3.3         </span></span>
<span><span class="co">#&gt; [16] utf8_1.2.4           yaml_2.3.8           data.table_1.15.2   </span></span>
<span><span class="co">#&gt; [19] knitr_1.45           prettyunits_1.2.0    labeling_0.4.3      </span></span>
<span><span class="co">#&gt; [22] htmlwidgets_1.6.4    scatterplot3d_0.3-44 RColorBrewer_1.1-3  </span></span>
<span><span class="co">#&gt; [25] Rtsne_0.17           withr_3.0.0          purrr_1.0.2         </span></span>
<span><span class="co">#&gt; [28] desc_1.4.3           grid_4.3.3           fansi_1.0.6         </span></span>
<span><span class="co">#&gt; [31] fastTopics_0.6-184   colorspace_2.1-0     scales_1.3.0        </span></span>
<span><span class="co">#&gt; [34] gtools_3.9.5         cli_3.6.2            rmarkdown_2.26      </span></span>
<span><span class="co">#&gt; [37] crayon_1.5.2         ragg_1.2.7           generics_0.1.3      </span></span>
<span><span class="co">#&gt; [40] RcppParallel_5.1.7   httr_1.4.7           pbapply_1.7-2       </span></span>
<span><span class="co">#&gt; [43] cachem_1.0.8         splines_4.3.3        parallel_4.3.3      </span></span>
<span><span class="co">#&gt; [46] softImpute_1.4-1     vctrs_0.6.5          Matrix_1.6-5        </span></span>
<span><span class="co">#&gt; [49] jsonlite_1.8.8       hms_1.1.3            mixsqp_0.3-54       </span></span>
<span><span class="co">#&gt; [52] ggrepel_0.9.5        irlba_2.3.5.1        horseshoe_0.2.0     </span></span>
<span><span class="co">#&gt; [55] systemfonts_1.0.6    trust_0.1-8          plotly_4.10.4       </span></span>
<span><span class="co">#&gt; [58] jquerylib_0.1.4      tidyr_1.3.1          glue_1.7.0          </span></span>
<span><span class="co">#&gt; [61] pkgdown_2.0.7        uwot_0.1.16          Polychrome_1.5.1    </span></span>
<span><span class="co">#&gt; [64] gtable_0.3.4         quadprog_1.5-8       munsell_0.5.0       </span></span>
<span><span class="co">#&gt; [67] tibble_3.2.1         pillar_1.9.0         htmltools_0.5.7     </span></span>
<span><span class="co">#&gt; [70] truncnorm_1.0-9      R6_2.5.1             textshaping_0.3.7   </span></span>
<span><span class="co">#&gt; [73] evaluate_0.23        lattice_0.22-5       highr_0.10          </span></span>
<span><span class="co">#&gt; [76] RhpcBLASctl_0.23-42  memoise_2.0.1        SQUAREM_2021.1      </span></span>
<span><span class="co">#&gt; [79] ashr_2.2-66          bslib_0.6.1          Rcpp_1.0.12         </span></span>
<span><span class="co">#&gt; [82] deconvolveR_1.2-1    xfun_0.42            fs_1.6.3            </span></span>
<span><span class="co">#&gt; [85] pkgconfig_2.0.3</span></span></code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Jason Willwerscheid, Peter Carbonetto, Wei Wang, Matthew Stephens.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
