<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Introduction to flashier • flashier</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Introduction to flashier">
<meta property="og:description" content="flashier">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">flashier</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.0.53</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">Home</a>
</li>
<li>
  <a href="../articles/index.html">Vignettes</a>
</li>
<li>
  <a href="../reference/index.html">Functions</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/willwerscheid/flashier" class="external-link">Source</a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Introduction to flashier</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/willwerscheid/flashier/blob/HEAD/vignettes/flashier_intro.Rmd" class="external-link"><code>vignettes/flashier_intro.Rmd</code></a></small>
      <div class="hidden name"><code>flashier_intro.Rmd</code></div>

    </div>

    
    
<p>The matrix we analyze in this example is a 1,000 x 44 matrix of
z-scores quantifying support for association between genetic variants in
the human genome (the rows of the matrix) and gene expression measured
across 44 human tissues (the columns):</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/willwerscheid/flashier" class="external-link">flashier</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">gtex</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">gtex</span><span class="op">)</span> </span>
<span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">gtex</span><span class="op">)</span></span>
<span><span class="va">gtex</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span></span>
<span><span class="co">#&gt; [1] 1000</span></span>
<span><span class="co">#&gt; [1] 44</span></span>
<span><span class="co">#&gt;                                       Adipose_Subcutaneous</span></span>
<span><span class="co">#&gt; ENSG00000099977.9_22_24266954_A_C_b37             8.099352</span></span>
<span><span class="co">#&gt; ENSG00000233868.1_2_222475922_A_G_b37             1.079264</span></span>
<span><span class="co">#&gt;                                       Adipose_Visceral_Omentum</span></span>
<span><span class="co">#&gt; ENSG00000099977.9_22_24266954_A_C_b37                6.4129683</span></span>
<span><span class="co">#&gt; ENSG00000233868.1_2_222475922_A_G_b37               -0.7760486</span></span></code></pre></div>
<p>The genetic variants are linked to genes, so each row of the matrix
corresponds to a gene (the names of the rows are the genes’ Ensembl
ids). These z-scores were originally calculated in <span class="citation">Urbut et al. (2019)</span> using data made available
from the Genotype Tissue Expression (GTEx) project <span class="citation">(Lonsdale et al. 2013)</span>, so we refer to this as
the “GTEx data set.”</p>
<p>Our motivation for factorizing the GTEx data matrix is to identify
patterns (“factors”) that give insight into the genetics of the 44
tissues: in particular, which tissues tend to act in concert, and to
what degree? First, to get some intuition into the structure of this
data set, we visualize the tissues in a 2-d embedding generated by <span class="math inline">\(t\)</span>-SNE <span class="citation">(L. Van der
Maaten and Hinton 2008; L. J. P. Van der Maaten 2014; Krijthe
2015)</span>:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/jkrijthe/Rtsne" class="external-link">Rtsne</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggrepel.slowkow.com/" class="external-link">ggrepel</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://wilkelab.org/cowplot/" class="external-link">cowplot</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">out</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Rtsne/man/Rtsne.html" class="external-link">Rtsne</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">gtex</span><span class="op">)</span>, dims <span class="op">=</span> <span class="fl">2</span>, perplexity <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span>
<span><span class="va">pdat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>d1 <span class="op">=</span> <span class="va">out</span><span class="op">$</span><span class="va">Y</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span>,</span>
<span>                   d2 <span class="op">=</span> <span class="va">out</span><span class="op">$</span><span class="va">Y</span><span class="op">[</span>, <span class="fl">2</span><span class="op">]</span>,</span>
<span>                   tissue <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">gtex</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">pdat</span>,<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">d1</span>, y <span class="op">=</span> <span class="va">d2</span>, label <span class="op">=</span> <span class="va">tissue</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">3</span>, color <span class="op">=</span> <span class="va">gtex_colors</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggrepel.slowkow.com/reference/geom_text_repel.html" class="external-link">geom_text_repel</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">2</span>, max.overlaps <span class="op">=</span> <span class="cn">Inf</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"t-SNE [1]"</span>, y <span class="op">=</span> <span class="st">"t-SNE [2]"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_classic</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.ticks <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>        axis.text <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="flashier_intro_files/figure-html/tsne-1.png" width="480" style="display: block; margin: auto;"></p>
<p>Visually, one of the predominant trends is the clustering of brain
tissues reflecting the greater genetic similarity among brain tissues.
But there may be other interesting patterns that are revealed more
effectively by a matrix factorization.</p>
<p>The flashier package implements the <em>empirical Bayes matrix
factorization</em> (EBMF) framework developed by <span class="citation">W. Wang and Stephens (2021)</span>. EBMF is based on
the following model of an <span class="math inline">\(n \times
p\)</span> data matrix <span class="math inline">\({\bf X}\)</span>:</p>
<p><span class="math display">\[
\mathbf{X} = \mathbf{L} \mathbf{F}^\top + \mathbf{E} \\
e_{ij} \sim \mathcal{N}(0, \sigma^2) \\
\ell_{ik} \sim g_\ell^{(k)} \in \mathcal{G}_{\ell} \\
f_{jk} \sim g_f^{(k)} \in \mathcal{G}_f,
\]</span></p>
<p>where <span class="math inline">\({\bf L}, {\bf F}, {\bf E}\)</span>
are, respectively, matrices of dimension <span class="math inline">\(n
\times K\)</span>, <span class="math inline">\(p \times K\)</span> and
<span class="math inline">\(n \times p\)</span> storing real-valued
elements <span class="math inline">\(l_{ik}\)</span>, <span class="math inline">\(f_{jk}\)</span>, and <span class="math inline">\(e_{ij}\)</span>. The matrices <span class="math inline">\({\bf  L}\)</span> and <span class="math inline">\({\bf F}\)</span> form an <em>approximate low-rank
representation</em> of <span class="math inline">\({\bf X}\)</span>,
<span class="math inline">\({\bf X} \approx {\bf L} {\bf
F}^\top\)</span>, and this low-rank representation is learned from the
data. <span class="math inline">\(K\)</span> specifies the rank of the
reduced representation, and is typically set to a positive number much
smaller than <span class="math inline">\(n\)</span> and <span class="math inline">\(p\)</span>.</p>
<p>The flexibility of the EBMF framework comes from its ability to
accommodate, in principle, any choice of prior family for <span class="math inline">\(\mathcal{G}_{\ell}\)</span> and <span class="math inline">\(\mathcal{G}_f\)</span>. As we illustrate below,
different choices of prior families can lead to very different
factorizations. Some choices closely correspond to existing matrix
factorization methods. For example, if <span class="math inline">\(\mathcal{G}_{\ell}\)</span> and <span class="math inline">\(\mathcal{G}_f\)</span> are both the families of
zero-centered normal priors, then the low-rank representation <span class="math inline">\({\bf L}{\bf F}^\top\)</span> is expected to be
similar to a truncated singular value decomposition (SVD) <span class="citation">(Nakajima and Sugiyama 2011)</span>. When point-normal
prior families are used instead, one obtains empirical Bayes versions of
sparse SVD or sparse factor analysis <span class="citation">(Engelhardt
2010; Yang, Ma, and Buja 2014; Witten, Tibshirani, and Hastie
2009)</span>. Thus, EBMF is a highly flexible framework for matrix
factorization that includes important previous methods as special cases,
but also many new combinations.</p>
<p>Since fitting the EBMF model can be reduced to solving a sequence of
EBNM problems <span class="citation">(W. Wang and Stephens 2021)</span>,
flashier leverages ebnm for the core model fitting routines. In brief,
fitting an EBMF model involves solving an EBNM problem separately for
each column of <span class="math inline">\({\bf L}\)</span> and for each
column of <span class="math inline">\({\bf F}\)</span>; the former
results in a fitted prior <span class="math inline">\(\hat{g}_{\ell}^{(k)}\)</span> and posterior
estimates of entries <span class="math inline">\(\ell_{ik}\)</span>, and
the latter results in a fitted prior <span class="math inline">\(\hat{g}_f^{(k)}\)</span> and posteror estimates of
entries <span class="math inline">\(f_{jk}\)</span>. These EBNM models
are repeatedly re-fitted until the priors and posterior estimates
converge to a fixed point. Therefore, fitting an EBMF model can involve
solving many hundreds or even many thousands of EBNM problems.
Leveraging the efficient EBNM solvers available in ebnm makes it
possible for flashier to tackle large-scale matrix factorization
problems.</p>
<p>The flashier function <code><a href="../reference/flash.html">flash()</a></code> provides the main
interface for fitting EBMF models. The following call to
<code><a href="../reference/flash.html">flash()</a></code> fits a matrix factorization to the GTEx data with
normal priors on <span class="math inline">\({\bf L}\)</span> and <span class="math inline">\({\bf F}\)</span>:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">flash_n</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/flash.html">flash</a></span><span class="op">(</span><span class="va">gtex</span>, ebnm_fn <span class="op">=</span> <span class="va">ebnm_normal</span>, backfit <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt; Adding factor 1 to flash object...</span></span>
<span><span class="co">#&gt; Adding factor 2 to flash object...</span></span>
<span><span class="co">#&gt; Adding factor 3 to flash object...</span></span>
<span><span class="co">#&gt; Adding factor 4 to flash object...</span></span>
<span><span class="co">#&gt; Adding factor 5 to flash object...</span></span>
<span><span class="co">#&gt; Adding factor 6 to flash object...</span></span>
<span><span class="co">#&gt; Adding factor 7 to flash object...</span></span>
<span><span class="co">#&gt; Adding factor 8 to flash object...</span></span>
<span><span class="co">#&gt; Adding factor 9 to flash object...</span></span>
<span><span class="co">#&gt; Adding factor 10 to flash object...</span></span>
<span><span class="co">#&gt; Adding factor 11 to flash object...</span></span>
<span><span class="co">#&gt; Adding factor 12 to flash object...</span></span>
<span><span class="co">#&gt; Adding factor 13 to flash object...</span></span>
<span><span class="co">#&gt; Adding factor 14 to flash object...</span></span>
<span><span class="co">#&gt; Adding factor 15 to flash object...</span></span>
<span><span class="co">#&gt; Factor doesn't significantly increase objective and won't be added.</span></span>
<span><span class="co">#&gt; Wrapping up...</span></span>
<span><span class="co">#&gt; Done.</span></span>
<span><span class="co">#&gt; Backfitting 14 factors (tolerance: 6.56e-04)...</span></span>
<span><span class="co">#&gt;   Difference between iterations is within 1.0e+02...</span></span>
<span><span class="co">#&gt;   Difference between iterations is within 1.0e+01...</span></span>
<span><span class="co">#&gt;   Difference between iterations is within 1.0e+00...</span></span>
<span><span class="co">#&gt;   Difference between iterations is within 1.0e-01...</span></span>
<span><span class="co">#&gt;   Difference between iterations is within 1.0e-02...</span></span>
<span><span class="co">#&gt; Wrapping up...</span></span>
<span><span class="co">#&gt; Done.</span></span>
<span><span class="co">#&gt; Nullchecking 14 factors...</span></span>
<span><span class="co">#&gt; Done.</span></span></code></pre></div>
<p>The “ebnm_fn” argument specifies the prior family for <span class="math inline">\({\bf L}\)</span> and <span class="math inline">\({\bf F}\)</span>, and accepts any of the prior
family functions implemented in the ebnm package (e.g.,
<code>ebnm_point_normal</code> or <code>ebnm_unimodal</code>; see Table
X for a near-complete list of prior families). Specifying the number of
factors <span class="math inline">\(K\)</span> is not needed because
flashier automatically tries to determine an appropriate rank by
checking whether the addition of new factors substantially improves the
fit. (If one prefers a smaller number of factors than what is determined
automatically, the maximum number of factors can be adjusted with the
“greedy_Kmax” argument.) Here we also turned on backfitting by setting
<code>backfit = TRUE</code>. This is generally recommended because it
improves the quality of the fit <span class="citation">(W. Wang and
Stephens 2021)</span>, with the caveat that it may make the model
fitting too slow for very large data sets. Even with backfitting, the
full computation here is very fast; when running this example on a
current MacBook Pro, for example, the model fitting is completed in less
than one second.</p>
<p>The <code><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot()</a></code> method for flash objects can be used to
visualize the estimated matrix <span class="math inline">\({\bf
L}\)</span>. Adding color to distinguish among tissues yields an
effective visualization that aids interpretation of factors:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">flash_n</span>, pm_which <span class="op">=</span> <span class="st">"factors"</span>, pm_colors <span class="op">=</span> <span class="va">gtex_colors</span>, plot_type <span class="op">=</span> <span class="st">"bar"</span><span class="op">)</span></span></code></pre></div>
<p><img src="flashier_intro_files/figure-html/plot-flash-normal-1.png" width="600" style="display: block; margin: auto;"></p>
<p>The first factor sets the “baseline” z-score for each tissue. Other
factors appear to capture expected trends: for example, factor 2 picks
up a z-score pattern that is more dominant in brain tissues. More
tissue-specific patterns are picked up by factor 3, which captures whole
blood effects, and factor 6, which captures effects more specific to
testis.</p>
<p>SVD and SVD-like matrix factorizations may provide compact
approximations of a data matrix, but they are generally known to have
poor interpretability. “Sparse” factorizations can produce factors that
are much more individually interpretable. In the EBNM framework, a
sparse matrix factorization can be achieved by choosing a flexible prior
family that better encourages sparsity in <span class="math inline">\({\bf L}\)</span> and/or <span class="math inline">\({\bf F}\)</span>. The flashier interface makes it
straightforward to experiment with different prior families; for
example, we can use point-normal priors by simply modifying the
“ebnm_fn” argument:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">flash_pn</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/flash.html">flash</a></span><span class="op">(</span><span class="va">gtex</span>, ebnm_fn <span class="op">=</span> <span class="va">ebnm_point_normal</span>, backfit <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">flash_pn</span>, pm_which <span class="op">=</span> <span class="st">"factors"</span>, pm_colors <span class="op">=</span> <span class="va">gtex_colors</span>, plot_type <span class="op">=</span> <span class="st">"bar"</span><span class="op">)</span></span></code></pre></div>
<p><img src="flashier_intro_files/figure-html/flash-point-normal-1.png" width="720" style="display: block; margin: auto;"></p>
<pre><code><span><span class="co">#&gt; Adding factor 1 to flash object...</span></span>
<span><span class="co">#&gt; Adding factor 2 to flash object...</span></span>
<span><span class="co">#&gt; Adding factor 3 to flash object...</span></span>
<span><span class="co">#&gt; Adding factor 4 to flash object...</span></span>
<span><span class="co">#&gt; Adding factor 5 to flash object...</span></span>
<span><span class="co">#&gt; Adding factor 6 to flash object...</span></span>
<span><span class="co">#&gt; Adding factor 7 to flash object...</span></span>
<span><span class="co">#&gt; Adding factor 8 to flash object...</span></span>
<span><span class="co">#&gt; Adding factor 9 to flash object...</span></span>
<span><span class="co">#&gt; Adding factor 10 to flash object...</span></span>
<span><span class="co">#&gt; Adding factor 11 to flash object...</span></span>
<span><span class="co">#&gt; Adding factor 12 to flash object...</span></span>
<span><span class="co">#&gt; Adding factor 13 to flash object...</span></span>
<span><span class="co">#&gt; Adding factor 14 to flash object...</span></span>
<span><span class="co">#&gt; Adding factor 15 to flash object...</span></span>
<span><span class="co">#&gt; Adding factor 16 to flash object...</span></span>
<span><span class="co">#&gt; Adding factor 17 to flash object...</span></span>
<span><span class="co">#&gt; Adding factor 18 to flash object...</span></span>
<span><span class="co">#&gt; Adding factor 19 to flash object...</span></span>
<span><span class="co">#&gt; Adding factor 20 to flash object...</span></span>
<span><span class="co">#&gt; Adding factor 21 to flash object...</span></span>
<span><span class="co">#&gt; Adding factor 22 to flash object...</span></span>
<span><span class="co">#&gt; Factor doesn't significantly increase objective and won't be added.</span></span>
<span><span class="co">#&gt; Wrapping up...</span></span>
<span><span class="co">#&gt; Done.</span></span>
<span><span class="co">#&gt; Backfitting 21 factors (tolerance: 6.56e-04)...</span></span>
<span><span class="co">#&gt;   Difference between iterations is within 1.0e+02...</span></span>
<span><span class="co">#&gt;   Difference between iterations is within 1.0e+01...</span></span>
<span><span class="co">#&gt;   Difference between iterations is within 1.0e+00...</span></span>
<span><span class="co">#&gt;   Difference between iterations is within 1.0e-01...</span></span>
<span><span class="co">#&gt;   Difference between iterations is within 1.0e-02...</span></span>
<span><span class="co">#&gt;   Difference between iterations is within 1.0e-03...</span></span>
<span><span class="co">#&gt; Wrapping up...</span></span>
<span><span class="co">#&gt; Done.</span></span>
<span><span class="co">#&gt; Nullchecking 21 factors...</span></span>
<span><span class="co">#&gt; Done.</span></span></code></pre>
<p>The point-normal priors require estimation of slightly more
parameters than the normal priors, but the call to <code><a href="../reference/flash.html">flash()</a></code>
is still fast; it took less than 5 seconds to run on our MacBook
Pro.</p>
<p>As before, the first factor acts as a “baseline” and the second
factor captures z-scores largely specific to brain tissues. However,
there are also some major differences: some factors are much more
clearly tissue-specific (e.g., factors 3 and 4 for, respectively, whole
blood and testis); other patterns include cerebellar-specific patterns
(factor 9) and heart-specific patterns (factor 12). These same patterns
are also captured in the matrix factorization with normal priors, but
they often appear in combination with other patterns in unintuitive
ways.</p>
<p>Another matrix factorization approach that has been recently proposed
is <em>semi-nonnegative matrix factorization</em> <span class="citation">(Ding, Li, and Jordan 2010; M. Wang, Fischer, and Song
2019; He et al. 2020)</span>, in which the elements of either <span class="math inline">\({\bf L}\)</span> or <span class="math inline">\({\bf F}\)</span> (but not both) are constrained to
be nonnegative. Supposing that <span class="math inline">\({\bf
L}\)</span> is constrained to be nonnegative, the elements <span class="math inline">\(\ell_{ik} \geq 0\)</span> can be viewed as
“weights” or “memberships,” and each row of <span class="math inline">\({\bf X}\)</span> is then interpretable as a
weighted combination of factors: <span class="math inline">\(x_{i \cdot}
\approx \sum_{k=1}^K \ell_{ik} f_{\cdot k}\)</span>.</p>
<p>To obtain a semi-nonnegative matrix factorization via flashier, one
need only choose appropriate prior families. Here we use point-normal
priors for <span class="math inline">\({\bf L}\)</span> and
point-exponential priors for <span class="math inline">\({\bf
F}\)</span>; thus we combine the benefits of both semi-nonnegative and
sparse matrix factorization. In flashier, we can specify different
priors for <span class="math inline">\({\bf L}\)</span> and <span class="math inline">\({\bf F}\)</span> by setting the “ebnm_fn” argument
to be a list in which the first element is the prior family for <span class="math inline">\({\bf L}\)</span> and the second the prior family
for <span class="math inline">\({\bf F}\)</span>:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">flash_snn</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/flash.html">flash</a></span><span class="op">(</span><span class="va">gtex</span>, </span>
<span>                   ebnm_fn <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">ebnm_point_normal</span>, <span class="va">ebnm_point_exponential</span><span class="op">)</span>, </span>
<span>                   backfit <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">flash_snn</span>, pm_which <span class="op">=</span> <span class="st">"factors"</span>, pm_colors <span class="op">=</span> <span class="va">gtex_colors</span>, plot_type <span class="op">=</span> <span class="st">"bar"</span><span class="op">)</span></span></code></pre></div>
<p><img src="flashier_intro_files/figure-html/flash-snn-1.png" width="720" style="display: block; margin: auto;"></p>
<pre><code><span><span class="co">#&gt; Adding factor 1 to flash object...</span></span>
<span><span class="co">#&gt; Adding factor 2 to flash object...</span></span>
<span><span class="co">#&gt; Adding factor 3 to flash object...</span></span>
<span><span class="co">#&gt; Adding factor 4 to flash object...</span></span>
<span><span class="co">#&gt; Adding factor 5 to flash object...</span></span>
<span><span class="co">#&gt; Adding factor 6 to flash object...</span></span>
<span><span class="co">#&gt; Adding factor 7 to flash object...</span></span>
<span><span class="co">#&gt; Adding factor 8 to flash object...</span></span>
<span><span class="co">#&gt; Adding factor 9 to flash object...</span></span>
<span><span class="co">#&gt; Adding factor 10 to flash object...</span></span>
<span><span class="co">#&gt; Adding factor 11 to flash object...</span></span>
<span><span class="co">#&gt; Adding factor 12 to flash object...</span></span>
<span><span class="co">#&gt; Adding factor 13 to flash object...</span></span>
<span><span class="co">#&gt; Adding factor 14 to flash object...</span></span>
<span><span class="co">#&gt; Adding factor 15 to flash object...</span></span>
<span><span class="co">#&gt; Adding factor 16 to flash object...</span></span>
<span><span class="co">#&gt; Adding factor 17 to flash object...</span></span>
<span><span class="co">#&gt; Adding factor 18 to flash object...</span></span>
<span><span class="co">#&gt; Adding factor 19 to flash object...</span></span>
<span><span class="co">#&gt; Adding factor 20 to flash object...</span></span>
<span><span class="co">#&gt; Adding factor 21 to flash object...</span></span>
<span><span class="co">#&gt; Adding factor 22 to flash object...</span></span>
<span><span class="co">#&gt; Factor doesn't significantly increase objective and won't be added.</span></span>
<span><span class="co">#&gt; Wrapping up...</span></span>
<span><span class="co">#&gt; Done.</span></span>
<span><span class="co">#&gt; Backfitting 21 factors (tolerance: 6.56e-04)...</span></span>
<span><span class="co">#&gt;   Difference between iterations is within 1.0e+02...</span></span>
<span><span class="co">#&gt;   Difference between iterations is within 1.0e+01...</span></span>
<span><span class="co">#&gt;   Difference between iterations is within 1.0e+00...</span></span>
<span><span class="co">#&gt;   Difference between iterations is within 1.0e-01...</span></span>
<span><span class="co">#&gt;   Difference between iterations is within 1.0e-02...</span></span>
<span><span class="co">#&gt;   Difference between iterations is within 1.0e-03...</span></span>
<span><span class="co">#&gt; Wrapping up...</span></span>
<span><span class="co">#&gt; Done.</span></span>
<span><span class="co">#&gt; Nullchecking 21 factors...</span></span>
<span><span class="co">#&gt; Done.</span></span></code></pre>
<p>This code returned results in less than 3 seconds on our MacBook
Pro.</p>
<p>The result is in many ways strikingly similar to the sparse
factorization obtained using point-normal priors. In some cases, the
same tissue-specific effects are recovered, up to a difference in signs
(e.g., factor 5 for fibroblasts and factor 8 for thyroid). But the
semi-nonnegative factorization identifies factors that are arguably more
intuitive; for example, point-normal factor 10, which combines artery
and adipose tissues, is split into two factors in the semi-nonnegative
factorization (factor 9 for artery and factor 11 for adipose
tissue).</p>
<p>The flashier package has many more options which we did not explore
here, including other choices of prior family, different options for
estimating the variances of the residuals <span class="math inline">\({\bf E}\)</span>, and options for fine-tuning the
model fitting to improve the speed and quality of EBMF model fits for
larger data sets. Many of these options are discussed in the other
vignettes.</p>
<div class="section level2">
<h2 id="session-information">Session information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h2>
<p>The following R version and packages were used to generate this
vignette:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.3.3 (2024-02-29)</span></span>
<span><span class="co">#&gt; Platform: aarch64-apple-darwin20 (64-bit)</span></span>
<span><span class="co">#&gt; Running under: macOS Sonoma 14.5</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRblas.0.dylib </span></span>
<span><span class="co">#&gt; LAPACK: /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.11.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt; [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: America/Chicago</span></span>
<span><span class="co">#&gt; tzcode source: internal</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt; [1] cowplot_1.1.3   ggrepel_0.9.5   ggplot2_3.5.0   Rtsne_0.17     </span></span>
<span><span class="co">#&gt; [5] flashier_1.0.53 ebnm_1.1-34    </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;  [1] tidyselect_1.2.1     viridisLite_0.4.2    dplyr_1.1.4         </span></span>
<span><span class="co">#&gt;  [4] farver_2.1.1         fastmap_1.1.1        lazyeval_0.2.2      </span></span>
<span><span class="co">#&gt;  [7] digest_0.6.34        lifecycle_1.0.4      invgamma_1.1        </span></span>
<span><span class="co">#&gt; [10] magrittr_2.0.3       compiler_4.3.3       rlang_1.1.3         </span></span>
<span><span class="co">#&gt; [13] sass_0.4.8           progress_1.2.3       tools_4.3.3         </span></span>
<span><span class="co">#&gt; [16] utf8_1.2.4           yaml_2.3.8           data.table_1.15.2   </span></span>
<span><span class="co">#&gt; [19] knitr_1.45           prettyunits_1.2.0    labeling_0.4.3      </span></span>
<span><span class="co">#&gt; [22] htmlwidgets_1.6.4    scatterplot3d_0.3-44 RColorBrewer_1.1-3  </span></span>
<span><span class="co">#&gt; [25] withr_3.0.0          purrr_1.0.2          desc_1.4.3          </span></span>
<span><span class="co">#&gt; [28] grid_4.3.3           fansi_1.0.6          fastTopics_0.6-184  </span></span>
<span><span class="co">#&gt; [31] colorspace_2.1-0     scales_1.3.0         gtools_3.9.5        </span></span>
<span><span class="co">#&gt; [34] cli_3.6.2            rmarkdown_2.26       crayon_1.5.2        </span></span>
<span><span class="co">#&gt; [37] ragg_1.2.7           generics_0.1.3       RcppParallel_5.1.7  </span></span>
<span><span class="co">#&gt; [40] httr_1.4.7           pbapply_1.7-2        cachem_1.0.8        </span></span>
<span><span class="co">#&gt; [43] splines_4.3.3        parallel_4.3.3       softImpute_1.4-1    </span></span>
<span><span class="co">#&gt; [46] vctrs_0.6.5          Matrix_1.6-5         jsonlite_1.8.8      </span></span>
<span><span class="co">#&gt; [49] hms_1.1.3            mixsqp_0.3-54        irlba_2.3.5.1       </span></span>
<span><span class="co">#&gt; [52] horseshoe_0.2.0      systemfonts_1.0.6    trust_0.1-8         </span></span>
<span><span class="co">#&gt; [55] plotly_4.10.4        jquerylib_0.1.4      tidyr_1.3.1         </span></span>
<span><span class="co">#&gt; [58] glue_1.7.0           pkgdown_2.0.7        uwot_0.1.16         </span></span>
<span><span class="co">#&gt; [61] Polychrome_1.5.1     gtable_0.3.4         quadprog_1.5-8      </span></span>
<span><span class="co">#&gt; [64] munsell_0.5.0        tibble_3.2.1         pillar_1.9.0        </span></span>
<span><span class="co">#&gt; [67] htmltools_0.5.7      truncnorm_1.0-9      R6_2.5.1            </span></span>
<span><span class="co">#&gt; [70] textshaping_0.3.7    evaluate_0.23        lattice_0.22-5      </span></span>
<span><span class="co">#&gt; [73] highr_0.10           RhpcBLASctl_0.23-42  memoise_2.0.1       </span></span>
<span><span class="co">#&gt; [76] SQUAREM_2021.1       ashr_2.2-66          bslib_0.6.1         </span></span>
<span><span class="co">#&gt; [79] Rcpp_1.0.12          deconvolveR_1.2-1    xfun_0.42           </span></span>
<span><span class="co">#&gt; [82] fs_1.6.3             pkgconfig_2.0.3</span></span></code></pre></div>
</div>
<div class="section level2 unnumbered">
<h2 class="unnumbered" id="bibliography">Bibliography<a class="anchor" aria-label="anchor" href="#bibliography"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-DingJordan" class="csl-entry">
Ding, Chris H. Q., Tao Li, and Michael I. Jordan. 2010. <span>“Convex
and Semi-Nonnegative Matrix Factorizations.”</span> <em>IEEE
Transactions on Pattern Analysis and Machine Intelligence</em> 32 (1):
45–55.
</div>
<div id="ref-Engelhardt" class="csl-entry">
Engelhardt, Matthew, Barbara E. AND Stephens. 2010. <span>“Analysis of
Population Structure: A Unifying Framework and Novel Methods Based on
Sparse Factor Analysis.”</span> <em>PLOS Genetics</em> 6 (9): 1–12.
</div>
<div id="ref-he2020sn" class="csl-entry">
He, Yuan, Surya B. Chhetri, Marios Arvanitis, Kaushik Srinivasan,
François Aguet, Kristin G. Ardlie, Alvaro N. Barbeira, et al. 2020.
<span>“<span class="nocase">sn-spMF</span>: Matrix Factorization Informs
Tissue-Specific Genetic Regulation of Gene Expression.”</span>
<em>Genome Biology</em> 21: 235.
</div>
<div id="ref-rtsne" class="csl-entry">
Krijthe, Jesse H. 2015. <em>: <span class="math inline">\(t\)</span>-Distributed Stochastic Neighbor
Embedding Using a <span>Barnes-Hut</span> Implementation</em>. <a href="https://github.com/jkrijthe/Rtsne" class="external-link">https://github.com/jkrijthe/Rtsne</a>.
</div>
<div id="ref-GTEx" class="csl-entry">
Lonsdale, John, Jeffrey Thomas, Mike Salvatore, et al. 2013. <span>“The
Genotype-Tissue Expression (<span>GTE</span>x) Project.”</span>
<em>Nature Genetics</em> 45 (6): 580–85.
</div>
<div id="ref-Nakajima" class="csl-entry">
Nakajima, Shinichi, and Masashi Sugiyama. 2011. <span>“Theoretical
Analysis of <span>B</span>ayesian Matrix Factorization.”</span>
<em>Journal of Machine Learning Research</em> 12: 2583–2648.
</div>
<div id="ref-Urbut" class="csl-entry">
Urbut, Sarah M., Gao Wang, Peter Carbonetto, and Matthew Stephens. 2019.
<span>“Flexible Statistical Methods for Estimating and Testing Effects
in Genomic Studies with Multiple Conditions.”</span> <em>Nature
Genetics</em> 51: 187–95.
</div>
<div id="ref-tsne-2015" class="csl-entry">
Van der Maaten, L. J. P. 2014. <span>“Accelerating <span class="math inline">\(t\)</span>-<span>SNE</span> Using Tree-Based
Algorithms.”</span> <em>Journal of Machine Learning Research</em> 15:
3221–45.
</div>
<div id="ref-tSNE" class="csl-entry">
Van der Maaten, Laurens, and Geoffrey Hinton. 2008. <span>“Visualizing
Data Using <span class="math inline">\(t\)</span>-<span>SNE</span>.”</span> <em>Journal
of Machine Learning Research</em> 9 (11): 2579–2605.
</div>
<div id="ref-wang2019three" class="csl-entry">
Wang, Miaoyan, Jonathan Fischer, and Yun S Song. 2019. <span>“Three-Way
Clustering of Multi-Tissue Multi-Individual Gene Expression Data Using
Semi-Nonnegative Tensor Decomposition.”</span> <em>The Annals of Applied
Statistics</em> 13 (2): 1103–27.
</div>
<div id="ref-WangStephens" class="csl-entry">
Wang, Wei, and Matthew Stephens. 2021. <span>“Empirical
<span>Bayes</span> Matrix Factorization.”</span> <em>Journal of Machine
Learning Research</em> 22 (120): 1–40.
</div>
<div id="ref-Witten" class="csl-entry">
Witten, Daniela M, Robert Tibshirani, and Trevor Hastie. 2009. <span>“A
Penalized Matrix Decomposition, with Applications to Sparse Principal
Components and Canonical Correlation Analysis.”</span>
<em>Biostatistics</em> 10 (3): 515–34.
</div>
<div id="ref-SSVD" class="csl-entry">
Yang, Dan, Zongming Ma, and Andreas Buja. 2014. <span>“A Sparse Singular
Value Decomposition Method for High-Dimensional Data.”</span>
<em>Journal of Computational and Graphical Statistics</em> 23 (4):
923–42.
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Jason Willwerscheid, Peter Carbonetto, Wei Wang, Matthew Stephens.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
